--- gtk+-1.2.10.orig/configure.in	Fri Nov 15 18:14:52 2002
+++ gtk+-1.2.10/configure.in	Fri Nov 15 18:47:26 2002
@@ -184,7 +184,7 @@
 
 # i18n stuff
 ALL_LINGUAS="az ca cs da de el es et eu fi fr ga gl hr hu it ja ko lt nl no nn pl pt pt_BR ro ru sk sl sp sr sv tr uk vi wa zh_CN.GB2312 zh_TW.Big5"
-AM_GTK_GNU_GETTEXT
+AM_GNU_GETTEXT([external])
 LIBS="$LIBS $INTLLIBS"
 
 # AM_GTK_GNU_GETTEXT above substs $DATADIRNAME
