diff -urN gtk+-1.2.9.org/gdk/Makefile.am gtk+-1.2.9/gdk/Makefile.am
--- gtk+-1.2.9.org/gdk/Makefile.am	Sat Mar  3 15:05:45 2001
+++ gtk+-1.2.9/gdk/Makefile.am	Sat Mar  3 15:23:29 2001
@@ -41,7 +41,7 @@
 	gdktypes.h	\
 	gdkx.h		\
 @STRIP_END@
-gdk_c_sources = @STRIP_BEGIN@ \
+gdk_c_sources = \
 	gdk.c		\
 	gdkcc.c		\
 	gdkcolor.c	\
@@ -72,8 +72,7 @@
 	MwmUtil.h       \
 	gxid_lib.h	\
 	gxid_proto.h	\
-        gxid_lib.c	\
-@STRIP_END@
+        gxid_lib.c
 
 #
 # setup GDK sources and their dependancies
diff -urN gtk+-1.2.9.org/gtk/Makefile.am gtk+-1.2.9/gtk/Makefile.am
--- gtk+-1.2.9.org/gtk/Makefile.am	Sat Mar  3 15:05:45 2001
+++ gtk+-1.2.9/gtk/Makefile.am	Sat Mar  3 15:23:48 2001
@@ -154,7 +154,7 @@
 gtk_private_h_sources = @STRIP_BEGIN@ \
 @STRIP_END@
 # GTK+ C sources to build the library from
-gtk_c_sources = @STRIP_BEGIN@ \
+gtk_c_sources = \
 	gtkaccelgroup.c		\
 	gtkaccellabel.c		\
 	gtkadjustment.c		\
@@ -262,8 +262,7 @@
 	gtkwidget.c		\
 	gtkwindow.c		\
 	fnmatch.c		\
-	fnmatch.h		\
-@STRIP_END@
+	fnmatch.h
 # we use our own built_sources variable rules to avoid automake's
 # BUILT_SOURCES oddities
 # we generate frequently rebuild files piggyback on a stamp file, so sources
